pkg/rebalancer/storage/controller/service/rebalancer.go

```diff
@@ -19,28 +19,549 @@ package service
+	"github.com/vdaas/vald/internal/safety"
+	"github.com/vdaas/vald/pkg/rebalancer/storage/controller/config"
+	"github.com/vdaas/vald/pkg/rebalancer/storage/controller/model"
 )
 
+const (
+	jobType = "rebalancer"
+)
+
+// Rebalancer represents the rebalancer interface.
 type Rebalancer interface {
-	Start(context.Context) (<-chan error, error)
+	Start(ctx context.Context) (<-chan error, error)
 }
 
-type rebalancer struct{}
+type rebalancer struct {
+	podName      string
+	podNamespace string
+
+	jobs           atomic.Value
+	jobName        string
+	jobNamespace   string
+	jobTemplate    atomic.Value
+	jobTemplateKey string // config map key of job template
+
+	agentName         string
+	agentNamespace    string
+	agentResourceType config.AgentResourceType
+	pods              atomic.Value
+	podMetrics        atomic.Value
+
+	jobConfigmapName      string
+	jobConfigmapNamespace string
+
+	statefulSets atomic.Value
+
+	leaderElectionID string
 
-func New(opts ...Option) (dsc Rebalancer, err error) {
+	rcd           time.Duration // reconcile check duration
+	eg            errgroup.Group
+	ctrl          k8s.Controller
+	tolerance     float64
+	rateThreshold float64
+	decoder       *decoder.Decoder
+}
+
+// NewRebalancer initialize job, configmap, pod, podMetrics, statefulset, replicaset, daemonset reconciler.
+// And it returns the rebalancer implemenation or any error occurred.
+func NewRebalancer(opts ...RebalancerOption) (Rebalancer, error) {
 	r := new(rebalancer)
-	for _, opt := range append(defaultOpts, opts...) {
+
+	for _, opt := range append(defaultRebalancerOpts, opts...) {
 		if err := opt(r); err != nil {
 			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
 		}
 	}
+
+	job, err := job.New(
+		job.WithControllerName("job rebalancer"),
+		job.WithNamespaces(r.jobNamespace),
+		job.WithOnErrorFunc(func(err error) {
+			log.Error(err)
+		}),
+		job.WithOnReconcileFunc(func(jobList map[string][]job.Job) {
+			log.Debugf("[reconcile Job] length Joblist: %d", len(jobList))
+			jobs, ok := jobList[r.jobName]
+			if ok {
+				r.jobs.Store(jobs)
+			} else {
+				r.jobs.Store(make([]job.Job, 0))
+				log.Infof("job not found: %s", r.jobName)
+			}
+		}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	cm, err := configmap.New(
+		configmap.WithControllerName("configmap rebalancer"),
+		configmap.WithNamespaces(r.jobConfigmapNamespace),
+		configmap.WithOnErrorFunc(func(err error) {
+			log.Error(err)
+		}),
+		configmap.WithOnReconcileFunc(func(configmapList map[string][]configmap.ConfigMap) {
+			configmaps, ok := configmapList[r.jobConfigmapNamespace]
+			if ok {
+				for _, cm := range configmaps {
+					if cm.Name == r.jobConfigmapName {
+						if tmpl, ok := cm.Data[r.jobTemplateKey]; ok {
+							r.jobTemplate.Store(tmpl)
+						} else {
+							log.Infof("job template not found: %s", r.jobTemplateKey)
+						}
+						break
+					}
+				}
+			} else {
+				log.Infof("configmap not found: %s", r.jobConfigmapName)
+			}
+		}),
```

Comment:
if I made a misstake, I am sorry.
If we set `client.MatchingFields` to configmap option and use it when we call api, I think this logic become more simple 🤔 
but I have to fix [k8s / job](https://github.com/vdaas/vald/blob/a5999150f11c8130af8f2f9fbbcede7de7538314/internal/k8s/configmap/configmap.go#L74-L80) logic again.

The following is my image to use client option
```go                                             
opts := []client.Option {
    client.MatchingFields{
      "metadata.name": "jobConfigmapName",
    },
}                                    
```

Reply:
thanks you for your comment.
It has the place for improving the code and I know It would be better to implement more as we can.
But IMO I do not have to do in this phase because getting no review and no release.
How about implement in refactoring phase after complete basic rebalance component?

Reply:
sure, I agree with you, it looks good to me 👍